sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileIndex.scala

```diff
@@ -23,11 +23,30 @@ import org.apache.spark.sql.catalyst.InternalRow
 import org.apache.spark.sql.catalyst.expressions._
 import org.apache.spark.sql.types.StructType
 
+/**
+ * A file status augmented with optional metadata. File formats can use the extra metadata to expose
+ * custom file-constant metadata columns, but in general tasks and readers can use the per-file
+ * metadata however they see fit.
+ */
+case class FileStatusWithMetadata(fileStatus: FileStatus, metadata: Map[String, Any] = Map.empty) {
```

Comment:
Let's think more about the API design. I think it's too fragile to use `Any` in the API, without a well-defined rule for what the actually allowed values are.

I'd suggest using `Map[String, Literal]`. Then we can remove `def isSupportedType` as all types can be supported.

Reply:
See my [TODO](https://github.com/apache/spark/pull/40677/files/6d35127b60f77475bf4b158b762468f30ec3dd9a#diff-4445cc3828e35092eb261467b499b8b0ef69ae694ea8ce25abf16b8ef4b72fbaR282) above... we may need to consider supporting value-producing functions, to allow full pruning in cases where the value is somehow expensive to compute. Requiring `Literal` would block that (and AFAIK only `Any` could capture both `Literal` and `() => Literal`).

The `FILE_PATH` case that calls `Path.toString`, and the call sites of PartitionedFile is a small example of that possibility that got me thinking -- what if instead of passing length, path, etc as arguments, we just passed the actual file status, and used the extractors on it? Probably doesn't make sense to _actually_ do that for the hard-wired cases, tho.

Reply:
I do like the idea of supporting `Literal` as one of the supported cases -- it simplifies the type checking a bit, in that the "supported" primitive types are merely those for which the implementation will automatically create the `Literal` wrapper as a courtesy (similar to string vs. UTF8String).

Reply:
Update: I remember now another reason why I had added `isSupportedDataType` -- `ConstantColumnVector` (needed by [FileScanRDD...createMetadataColumnVector](https://github.com/apache/spark/pull/40677/files/6d35127b60f77475bf4b158b762468f30ec3dd9a#diff-b7b097f8cec6a7ae6640f9ecd6d4ac14ed304ad7c8db802a3ec2c0983535e157R200) below) supports a limited subset of types, and relies on type-specific getters and setters. Even if I wrote the (complex recursive) code to handle structs, maps, and arrays... we still wouldn't have complete coverage for all types.

Do we know for certain that `ConstantColumnVector` supports all types that can ever be encountered during vectorized execution? If not, we must keep the `isSupportedDataType` method I introduced.

Reply:
Updated the doc comment here to explain that file-source metadata fields is only one possible usage for the extra file metadata (which is conceptually at a deeper layer than catalyst and `Literal`).

Also updated `isSupportedType` to explain why not all types are supported.

Relevant implementation details: 
1. It would take a lot of work to support all data types, regardless of whether we use `Literal` vs. `Any`.
2. We anyway end up wrapping the provided value in a call to `Literal(_)`, because doing so simplifies null handling by making null-because-missing equivalent to null-because-null. At that point, we get wrapping of primitive values "for free" if we happen to pass `Any` instead.